File: leetcode/0872. Leaf-Similar Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        t1 = list()
        t2 = list()
        self.get_leaf_value(root1, t1)
        self.get_leaf_value(root2, t2)
        
        if len(t1) != len(t2):
            return False 

        for i in range(len(t1)):
            if t1[i] != t2[i]:
                return False

        return True
        

    def get_leaf_value(self, root: Optional[TreeNode], lis: List[int]) -> None:
        if root is None:
            return 
        
        if root.left is None and root.right is None:
            lis.append(root.val)
         
        self.get_leaf_value(root.left, lis)
        self.get_leaf_value(root.right, lis)
    

File: leetcode/0872. Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> t1Leaf;
        vector<int> t2Leaf;
        getLeafValue(root1, t1Leaf);
        getLeafValue(root2, t2Leaf);

        if (t1Leaf.size() != t2Leaf.size()) {
            return false;
        }

        for (int i = 0; i < t2Leaf.size(); ++i) {
            if (t1Leaf[i] != t2Leaf[i]) {
                return false;
            }
        }
        return true;
    }

    void getLeafValue(TreeNode* root, vector<int>& vec) {
        if (root == nullptr)
            return;

        if (root->left == root->right) {
            vec.push_back(root->val);
        }
        getLeafValue(root->left, vec);
        getLeafValue(root->right, vec);
    }
};